vnext/Microsoft.ReactNative/ReactPropertyBag.idl

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.ReactNative {
+
+  // The delegate is used to create property value on-demand.
+  [webhosthidden]
+  delegate Object ReactCreatePropertyValue();
+
+  // Namespace for the property name.
+  // Use ReactPropertyNameTable.GetNamespace to get atomic property namespace for a string.
+  [webhosthidden]
+  interface IReactPropertyNamespace {
+    // Get String name representation of the property namespace.
+    String NamespaceName { get; };
+  }
+
+  // Name of a property.
+  // Use ReactPropertyNameTable.GetName to get atomic property name for a string.
+  [webhosthidden]
+  interface IReactPropertyName {
+    // The local property name String in context of the property namespace.
+    String LocalName { get; };
+
+    // The namespace the property name is defined in.
+    IReactPropertyNamespace Namespace { get; };
+  }
+
+  // The IPropertyBag provides a thread-safe property storage.
+  // Properties are identified by a propertyName.
+  // have their own set of properties not visible to other components.
+  // It is expected that there will be no direct use of this interface.
+  // All usage should happen through a strongly typed accessors.
+  [webhosthidden]
+  runtimeclass ReactPropertyBag {
+    // Construct new PropertyNameTable.
+    ReactPropertyBag();
+
+    // Return a global namespace that an empty name.
+    static IReactPropertyNamespace GlobalNamespace { get; };
+
+    // Get an atomic IReactPropertyNamespace for a provided namespaceName.
+    // The valid namespace name must be at least one character long.
+    // Otherwise, the function crashes.
+    // Consider to use module name as a namespace for module-specific properties.
+    static IReactPropertyNamespace GetNamespace(String namespaceName);
+
+    // Get atomic IReactPropertyName for the provided property name and property namespace.
+    // If ns is null, then it uses IReactPropertyNamespace.GlobalNamespace.
+    // The valid property name must be at least one character long.
+    // Otherwise, the function crashes.
+    static IReactPropertyName GetName(IReactPropertyNamespace ns, String localName);
+
+    // Get property value for the propertyName. It returns null if the property does not exist.
```

Comment:
>returns null if the property does not exist [](start = 51, length = 43)

Does this mean that property bags cannot hold null values?

Reply:
Yes, given that SetProperty(..., null) is meant to remove a property.

Is this design intentional? IMHO, there sometimes is the need to differentiate between "a property is unknown" and "a property is null".

---
In reply to: [421738302](https://github.com/microsoft/react-native-windows/pull/4817#discussion_r421738302) [](ancestors = 421738302)

Reply:
Along the same lines: Why is there a 'RemoveProperty' method if a property can be removed with 'SetProperty(..., null)'? If the only answer is "convenience", I'd leave that to some platform/language projection while keeping the IDL minimally sufficient.

---
In reply to: [421740890](https://github.com/microsoft/react-native-windows/pull/4817#discussion_r421740890) [](ancestors = 421740890,421738302)

Reply:
Yes, I have added it for convenience. I really wanted it to have one method: Exchange to be more accurate what it does. IMHO, I want to have the behaviors similar to the std::atomic.
As for another question, I do not want to keep the null values. I think it makes the whole API more complex. The missing property returns today null. Technically, there is no observable difference if I remove property or keep it as null. In future I may a different internal data structure that may keep null values for sometime. I would rather consider it as an implementation detail.

Reply:
I'm ambivalent about this because I agree simplicity is good - on the other hand JS already has this notion through `undefined` vs. `null`... maybe it makes sense to match the JS paradigm and allow nulls as different from not set.

Reply:
I am reluctant doing it because it is not simple to implement and it will cause tons of confusion for users. It is quite painful for JavaScript developers. Many of them do not understand the difference. E.g. there are many talks where presenters propose as a good practice to use only one of them. We can always add it later, but for the last two years the PropertyBag existing in Office, I have never heard such request.